crates/engine/tree/src/tree/root.rs

```diff
@@ -1136,37 +1295,72 @@ where
     Ok(())
 }
 
-/// Returns accounts only with those storages that were not already fetched, and
-/// if there are no such storages and the account itself was already fetched, the
-/// account shouldn't be included.
-fn get_proof_targets(
-    state_update: &HashedPostState,
-    fetched_proof_targets: &MultiProofTargets,
-) -> MultiProofTargets {
-    let mut targets = MultiProofTargets::default();
-
-    // first collect all new accounts (not previously fetched)
-    for &hashed_address in state_update.accounts.keys() {
-        if !fetched_proof_targets.contains_key(&hashed_address) {
-            targets.insert(hashed_address, HashSet::default());
+/// Iterator that chunks proof targets.
+///
+/// This iterator will yield items of type [`Vec<MultiProofTargets>`], with each
+/// [`MultiProofTargets`] mapping having a maximum length of `accounts_chunk_size`, and each mapping
+/// value [`B256Set`] having a maximum length of `storages_chunk_size`.
+struct ChunkedProofTargets {
+    /// Sorted proof targets.
+    ///
+    /// We keep this sorted, because multiproofs are more efficient when proof targets are located
+    /// close to each other.
+    proof_targets: VecDeque<(B256, B256Set)>,
+    /// Number of accounts per chunk.
+    accounts_chunk_size: usize,
+    /// Number of storage slots per account per chunk.
+    storages_chunk_size: usize,
+}
+
+impl ChunkedProofTargets {
+    fn new(
+        proof_targets: MultiProofTargets,
+        accounts_chunk_size: usize,
+        storages_chunk_size: usize,
+    ) -> Self {
+        Self {
+            proof_targets: proof_targets
+                .into_iter()
+                .sorted_by_key(|(address, _)| *address)
+                .collect(),
+            accounts_chunk_size,
+            storages_chunk_size,
         }
     }
+}
+
+impl Iterator for ChunkedProofTargets {
+    type Item = Vec<MultiProofTargets>;
 
-    // then process storage slots for all accounts in the state update
-    for (hashed_address, storage) in &state_update.storages {
-        let fetched = fetched_proof_targets.get(hashed_address);
-        let mut changed_slots = storage
-            .storage
-            .keys()
-            .filter(|slot| !fetched.is_some_and(|f| f.contains(*slot)))
-            .peekable();
-
-        if changed_slots.peek().is_some() {
-            targets.entry(*hashed_address).or_default().extend(changed_slots);
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
this impl looks correct to me although unsure if this is the most optimal solution for chunking if we immediately consume the chunk as an iter.

don't have a better solution in mind though